active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,322 @@
+import Ember from "ember";
+
+// Usage: {{if-is-yes 'no'}}not rendered{{else}}rendered!{{/if-is-yes}}
+export default function ifIsYes(params, hash, blocks) {
+  if (params[0] === 'yes') {
+    blocks.template.yield();
+  } else {
+    blocks.inverse.yield();
+  }
+};
+```
+
+Porting this usage to a stateful helper only changes the boilerplate:
+
+```js
+// app/helpers/custom-if.js
+import Ember from "ember";
+
+// Usage: {{if-is-yes 'no'}}not rendered{{else}}rendered!{{/if-is-yes}}
+export default Ember.Helper.extend({
+  compute(params, hash, blocks) {
+    if (params[0]) {
+      blocks.template.yield();
+    } else {
+      blocks.inverse.yield();
+    }
+  }
+});
+```
+
+### Consuming services and recompute
+
+Stateful helpers are a valid target for service injection. For example:
+
+```js
+// app/helpers/current-user-name.js
+import Ember from "ember";
+
+export default Ember.Helper.extend({
+  // Same API as components:
+  session: Ember.inject.service(),
+  compute() {
+    return this.get('session.currentUser.name');
+  }
+});
+```
+
+However consuming a property from a service does not bind the data being
+displayed to that property. After `{{current-user-name}}` has been computed
+and rendered, it will never be invalidated.
+
+For this reason, stateful helpers are granted some control over their
+computation lifecycle. A stateful helper will recompute when:
+
+* A value passed via the template changes (`params` or `hash`)
+* The `recompute` method is called
+* Any key in the `recomputeObserves` array changes
+
+For example, this stateful helper checks if the current use has access to a
+resource type:
+
+```js
+// app/helpers/can-access.js
+import Ember from "ember";
+
+// Usage {{if (can-access 'admin') 'Welcome, boss' 'Heck no!'}}
+export default Ember.Helper.extend({
+  session: Ember.inject.service(),
+  recomputeObserves: ['session.currentUser'],
+  compute(params) {
+    const currentUser = this.get('session.currentUser');
+    return currentUser.can(params[0]);
+  }
+});
+```
+
+`recomputeObserves` can be desugared to:
+
+```js
+  recomputeWhen: Ember.observes('session.currentUser', function() {
+    this.recompute();
+  }),
+```
+
+# Drawbacks
+
+Stateful helpers may superficially appear similar to components, but in
+practice they have none of the special behavior of components. The intent of
+this RFC is that they remain conceptually very close to function helpers,
+but despite this intent they are a new concept for the framework.
+
+# Alternatives
+
+A [previous RFC](https://github.com/emberjs/rfcs/pull/52) explored creating a new class called Expressions, which would have
+more closely modeled the API of components (using positional params, attrs).
+After discussion and consideration it was clear that a third kind of template
+API would be very challenging to document and teach well.
+
+# Unresolved questions
+
+Helpers are one-way. Is it possible to have a two-way helper (data
+flowing upward like a mut)?
```

Comment:
I don't think we need to explore this in the RFC and I will remove the note. The idea would be to define semantics for:

```
<my-component isAdmin={{mut (can-access 'admin')}}>
```

This is not immediately useful, to my imagination.